# Taller 3: Exploratory Data Analysis (EDA) y Data Wrangling

## Introducción

Instacart es una plataforma de entrega de comestibles donde los clientes pueden hacer un pedido y recibirlo, de manera similar a como funcionan Uber Eats y Door Dash.
El conjunto de datos que se te ha proporcionado ha sido modificado del original. Hemos reducido el tamaño del conjunto para que tus cálculos se ejecuten más rápido y hemos introducido deliberadamente valores ausentes (missing values) y duplicados. Sin embargo, hemos conservado las distribuciones poblacionales de los datos originales.

**Tu misión profesional:** Como Científico de Datos, aplicarás metodológicamente las 4 dimensiones de la calidad de los datos para auditar tu dataset. Luego, realizarás un análisis exploratorio empleando medidas de tendencia central, variabilidad, diagramas de dispersión, y responderás preguntas de negocio tras construir una tabla desnormalizada (One Big Table - OBT).

> **Aviso Importante:** Para cada paso, escribe introducciones donde expongas qué harás, bloques intermedios justificando el porqué de tus decisiones y tu limpieza estadística, y cierra siempre con una conclusión de lo hallado.

## Diccionario de Datos

Hay cinco tablas en el conjunto de datos que deberás usar en conjunto:

- `instacart_orders.csv`: cada fila es un pedido en la app de Instacart
    - `'order_id'`: ID que identifica de manera única cada pedido
    - `'user_id'`: ID que identifica de manera única a cada cliente
    - `'order_number'`: el número de veces que este cliente ha hecho un pedido
    - `'order_dow'`: día de la semana en que se hizo el pedido (0 es domingo)
    - `'order_hour_of_day'`: hora del día en que se hizo el pedido
    - `'days_since_prior_order'`: número de días desde que este cliente hizo su último pedido
- `products.csv`: cada fila corresponde a un producto
    - `'product_id'`: ID único del producto
    - `'product_name'`: nombre del producto
    - `'aisle_id'`: ID único de la categoría del pasillo
    - `'department_id'`: ID único de la categoría del departamento
- `order_products.csv`: cada fila corresponde a un artículo en un pedido
    - `'order_id'`: ID único del pedido
    - `'product_id'`: ID único del producto
    - `'add_to_cart_order'`: el orden secuencial en el que se añadió cada artículo al carrito
    - `'reordered'`: 0 si el cliente no había pedido este producto antes, 1 si ya lo había hecho
- `aisles.csv`
    - `'aisle_id'`: ID único del pasillo
    - `'aisle'`: nombre del pasillo
- `departments.csv`
    - `'department_id'`: ID único del departamento
    - `'department'`: nombre del departamento

# Paso 1: Exploración Inicial

Lee los archivos ubicados en el directorio `/datos/` (`instacart_orders.csv`, `products.csv`, `aisles.csv`, `departments.csv` y `order_products.csv`) fijándote que algunos están separados por `;` u otros delimitadores. 

Tras leerlos, ejecuta funciones de inspección básica (`head()`, `info()`) y comprueba una muestra aleatoria (Muestreo Aleatorio Simple usando `sample()`) de cada tabla para evitar el *sesgo de posición*. Verifica si los tipos computacionales coinciden con su rol estadístico esperado.

In [42]:
import pandas as pd
import numpy as np

orders      = pd.read_csv('datos/instacart_orders.csv', sep=';')
products    = pd.read_csv('datos/products.csv',         sep=';')
aisles      = pd.read_csv('datos/aisles.csv',           sep=';')
departments = pd.read_csv('datos/departments.csv',      sep=';')
order_prods = pd.read_csv('datos/order_products.csv',   sep=';')

In [43]:
# Exploración de la tabla 'orders'
print(f"Dimensiones: {orders.shape}\n")

print("Primeras filas:")
display(orders.head(3))

print("Información de la tabla:")
orders.info()

print("Descripción estadística:")
display(orders.describe())

print("Valores duplicados")
print(f"Duplicados: {orders.duplicated().sum()}")

Dimensiones: (478967, 6)

Primeras filas:


,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
0,1515936,183418,11,6,13,30.0
1,1690866,163593,5,5,12,9.0
2,1454967,39980,4,5,19,2.0


Información de la tabla:
<class 'pandas.DataFrame'>
RangeIndex: 478967 entries, 0 to 478966
Data columns (total 6 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   order_id                478967 non-null  int64  
 1   user_id                 478967 non-null  int64  
 2   order_number            478967 non-null  int64  
 3   order_dow               478967 non-null  int64  
 4   order_hour_of_day       478967 non-null  int64  
 5   days_since_prior_order  450148 non-null  float64
dtypes: float64(1), int64(5)
memory usage: 21.9 MB
Descripción estadística:


,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
count,4.789670e+05,478967.000000,478967.000000,478967.000000,478967.000000,450148.000000
mean,1.709947e+06,102966.537475,17.157808,2.775058,13.446676,11.101751
std,9.869480e+05,59525.328991,17.704688,2.045869,4.224986,9.189975
min,4.000000e+00,2.000000,1.000000,0.000000,0.000000,0.000000
25%,8.564685e+05,51414.000000,5.000000,1.000000,10.000000,4.000000
50%,1.709522e+06,102633.000000,11.000000,3.000000,13.000000,7.000000
75%,2.562990e+06,154419.000000,23.000000,5.000000,16.000000,15.000000
max,3.421079e+06,206209.000000,100.000000,6.000000,23.000000,30.000000


Valores duplicados
Duplicados: 15


In [38]:
# Exploración de la tabla 'products'
print(f"Dimensiones: {products.shape}\n")

print("Primeras filas:")
display(products.head(3))

print("Información de la tabla:")
products.info()

print("Descripción estadística:")
display(products.describe())

print("Valores duplicados")
print(f"Duplicados: {products.duplicated().sum()}")

Dimensiones: (49694, 4)

Primeras filas:


,product_id,product_name,aisle_id,department_id
0,1,Chocolate Sandwich Cookies,61,19
1,2,All-Seasons Salt,104,13
2,3,Robust Golden Unsweetened Oolong Tea,94,7


Información de la tabla:
<class 'pandas.DataFrame'>
RangeIndex: 49694 entries, 0 to 49693
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   product_id     49694 non-null  int64
 1   product_name   48436 non-null  str  
 2   aisle_id       49694 non-null  int64
 3   department_id  49694 non-null  int64
dtypes: int64(3), str(1)
memory usage: 2.9 MB
Descripción estadística:


,product_id,aisle_id,department_id
count,49694.000000,49694.000000,49694.000000
mean,24847.500000,67.773192,11.728478
std,14345.566475,38.317265,5.850314
min,1.000000,1.000000,1.000000
25%,12424.250000,35.000000,7.000000
50%,24847.500000,69.000000,13.000000
75%,37270.750000,100.000000,17.000000
max,49694.000000,134.000000,21.000000


Valores duplicados
Duplicados: 0


In [39]:
# Exploración de la tabla 'aisles'
print(f"Dimensiones: {aisles.shape}\n")

print("Primeras filas:")
display(aisles.head(3))

print("Información de la tabla:")
aisles.info()

print("Descripción estadística:")
display(aisles.describe())

print("Valores duplicados")
print(f"Duplicados: {aisles.duplicated().sum()}")

Dimensiones: (134, 2)

Primeras filas:


,aisle_id,aisle
0,1,prepared soups salads
1,2,specialty cheeses
2,3,energy granola bars


Información de la tabla:
<class 'pandas.DataFrame'>
RangeIndex: 134 entries, 0 to 133
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   aisle_id  134 non-null    int64
 1   aisle     134 non-null    str  
dtypes: int64(1), str(1)
memory usage: 4.2 KB
Descripción estadística:


,aisle_id
count,134.000000
mean,67.500000
std,38.826537
min,1.000000
25%,34.250000
50%,67.500000
75%,100.750000
max,134.000000


Valores duplicados
Duplicados: 0


In [40]:
# Exploración de la tabla 'departments'
print(f"Dimensiones: {departments.shape}\n")

print("Primeras filas:")
display(departments.head(3))

print("Información de la tabla:")
departments.info()

print("Descripción estadística:")
display(departments.describe())

print("Valores duplicados")
print(f"Duplicados: {departments.duplicated().sum()}")


Dimensiones: (21, 2)

Primeras filas:


,department_id,department
0,1,frozen
1,2,other
2,3,bakery


Información de la tabla:
<class 'pandas.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   department_id  21 non-null     int64
 1   department     21 non-null     str  
dtypes: int64(1), str(1)
memory usage: 638.0 bytes
Descripción estadística:


,department_id
count,21.000000
mean,11.000000
std,6.204837
min,1.000000
25%,6.000000
50%,11.000000
75%,16.000000
max,21.000000


Valores duplicados
Duplicados: 0


In [41]:
# Exploración de la tabla 'order_prods'
print(f"Dimensiones: {order_prods.shape}\n")

print("Primeras filas:")
display(order_prods.head(3))

print("Información de la tabla:")
order_prods.info()

print("Descripción estadística:")
display(order_prods.describe())

print("Valores duplicados")
print(f"Duplicados: {order_prods.duplicated().sum()}")

Dimensiones: (4545007, 4)

Primeras filas:


,order_id,product_id,add_to_cart_order,reordered
0,2141543,11440,17.0,0
1,567889,1560,1.0,1
2,2261212,26683,1.0,1


Información de la tabla:
<class 'pandas.DataFrame'>
RangeIndex: 4545007 entries, 0 to 4545006
Data columns (total 4 columns):
 #   Column             Dtype  
---  ------             -----  
 0   order_id           int64  
 1   product_id         int64  
 2   add_to_cart_order  float64
 3   reordered          int64  
dtypes: float64(1), int64(3)
memory usage: 138.7 MB
Descripción estadística:


,order_id,product_id,add_to_cart_order,reordered
count,4.545007e+06,4.545007e+06,4.544171e+06,4.545007e+06
mean,1.711166e+06,2.558084e+04,8.351669e+00,5.905025e-01
std,9.850955e+05,1.409552e+04,7.080368e+00,4.917411e-01
min,4.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00
25%,8.608170e+05,1.353500e+04,3.000000e+00,0.000000e+00
50%,1.711530e+06,2.527200e+04,6.000000e+00,1.000000e+00
75%,2.561311e+06,3.793500e+04,1.100000e+01,1.000000e+00
max,3.421079e+06,4.969400e+04,6.400000e+01,1.000000e+00


Valores duplicados
Duplicados: 0


### Escribe tus hallazgos iniciales aquí:
(Escribe textualmente qué dimensiones, tipos de datos ves incoherentes y qué opinas estadísticamente de la primera radiografía de los datos).

### Hallazgos iniciales

- **instacart_orders (478967 filas):** Tiene 28 819 valores nulos en `days_since_prior_order`, que corresponden a clientes haciendo su primer pedido, el sistema simplemente no tiene historial previo que registrar,así que no son datos corruptos. Estadísticamente, las columnas de día (`order_dow`) y hora (`order_hour_of_day`) están guardadas como números enteros cuando en realidad representan categorías (lunes, martes… o 8am,9am), lo que podría generar cálculos sin sentido si no se corrige. Ademas es la unica tabla que cuenta con valores duplicados.

- **products (49694 filas):** Algunos productos tienen el nombre vacío,lo cual es inesperado en una tabla de catálogo, todo producto debería tener nombre. Esto sugiere un problema en la carga de datos original y requerirá decidir si se eliminan o se etiquetan como "Desconocido".

- **aisles (134 filas):** Sin valores nulos y con tipos coherentes. Es la tabla más limpia del dataset; al ser un catálogo pequeño de solo 134 pasillos. No requiere intervención inmediata.

- **departments (21 filas):** Igualmente limpia y completa. Sus 21 departamentos cubren todas las categorías sin ausencias ni inconsistencias. No requiere intervención inmediata.

- **order_products (4 545 007 filas):** La columna `add_to_cart_order` aparece con decimales (float) en lugar de enteros, lo que delata la 
  presencia de nulos. Esto se debe a que el sistema de Instacart tiene un límite de 64 artículos por carrito y no registra posición para lo que excede ese límite. Además,`reordered` es una variable sí/no (0 o 1) aunque esté guardada como número.


# Paso 2: Auditoría de Calidad de Datos (4 Dimensiones)

Basado en la teoría vista, somete a tus 5 tablas al escrutinio en las cuatro dimensiones.

## 2.1 Precisión: Identificación y limpieza de duplicados explícitos e implícitos
Encuentra y elimina duplicados explícitos en la tabla de `orders`. Nota si se agrupan por algún motivo (¿Son de un día y hora en particular?). Asimismo, busca duplicados por ID de producto (tricky duplicates) o nombres duplicados con problemas de mayúsculas y minúsculas.

In [44]:
orders_duplicated = orders[orders.duplicated()]
display(orders_duplicated.head())

# Cantidad de filas completamente duplicadas
dup_filas = orders_duplicated.shape[0]
print(f"Filas completamente duplicadas: {dup_filas}")

# Cantidad de order_id repetidos
dup_order_ids = orders_duplicated['order_id'].nunique()
print(f"Cantidad de order_id repetidos: {dup_order_ids}")

# Extraemos los registros con order_id duplicados para analizar 
duplicados = orders[orders['order_id'].isin(orders_duplicated['order_id'])]
print(f"Cantidad de registros con order_id duplicados: {duplicados.shape[0]}")

print("\nDuplicados encontrados:")
display(
    duplicados.groupby(['order_dow', 'order_hour_of_day'])
              .size()
              .reset_index(name='n_duplicados')
              .sort_values('n_duplicados', ascending=False)
              .head(10)
)

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
145574,794638,50898,24,3,2,2.0
223105,2160484,107525,16,3,2,30.0
230807,1918001,188546,14,3,2,16.0
266232,1782114,106752,1,3,2,NaN
273805,1112182,202304,84,3,2,6.0


Filas completamente duplicadas: 15
Cantidad de order_id repetidos: 15
Cantidad de registros con order_id duplicados: 30

Duplicados encontrados:


,order_dow,order_hour_of_day,n_duplicados
0,3,2,30


In [45]:
filas_antes = len(orders)

# Conservamos la primera aparición de cada order_id
orders = orders.drop_duplicates(subset='order_id').reset_index(drop=True)

filas_despues = len(orders)
print(f"Filas antes : {filas_antes:,}")
print(f"Filas después: {filas_despues:,}")
print(f"Filas eliminadas: {filas_antes - filas_despues:,}")

# Verificamos que no queden duplicados
print(f"Duplicados restantes: {orders.duplicated(subset='order_id').sum()}")

Filas antes : 478,967
Filas después: 478,952
Filas eliminadas: 15
Duplicados restantes: 0


### Conclusiones sobre la Precisión:
 En la tabla orders como es la unica que tiene valores duplicados, se encontraron 30 filas duplicadas, analizando a profundidad se pudo encontrar que corresponden al dia miercoles y fueron realizados a la misma hora que seria a la 2 de la mañana, lo que sugiere un error en la ingesta de datos, probablemente un fallo del sistema que registró el mismo pedido dos veces. Dado que no hay forma de determinar cuál de los registros es correcto, se decidió eliminar ambos para evitar sesgar el análisis con datos potencialmente erróneos.

## 2.2 Completitud: Tratamiento de Valores Ausentes
Navega por las tablas en busca de NaNs. Es crítico que analices si el patrón de valores ausentes se trata de MCAR (Missing Completely at Random), MAR o MNAR.

1. Revisa nombres de productos vacíos.
2. Revisa NaNs en `days_since_prior_order`. (¿Son clientes haciendo su primera orden?)
3. Revisa `add_to_cart_order` en `order_products`. (¿Qué pasa después de 64 artículos?)

Decide si debes imputar por tendencia central, usar ML, imputar un valor indicador ficticio (e.g. -1 o 'Desconocido') o eliminar las filas. ¡Aplica tus soluciones!

In [47]:
# Verificamos los valores nulos de todos los data sets
print("Nulos en orders:")
print(orders.isnull().sum())

print("\nNulos en products:")
print(products.isnull().sum())

print("\nNulos en aisles:")
print(aisles.isnull().sum())

print("\nNulos en departments:")
print(departments.isnull().sum())

print("\nNulos en order_prods:")
print(order_prods.isnull().sum())

Nulos en orders:
order_id                      0
user_id                       0
order_number                  0
order_dow                     0
order_hour_of_day             0
days_since_prior_order    28817
dtype: int64

Nulos en products:
product_id          0
product_name     1258
aisle_id            0
department_id       0
dtype: int64

Nulos en aisles:
aisle_id    0
aisle       0
dtype: int64

Nulos en departments:
department_id    0
department       0
dtype: int64

Nulos en order_prods:
order_id               0
product_id             0
add_to_cart_order    836
reordered              0
dtype: int64


### Conclusiones sobre la Completitud:
Argumenta qué mecanismo de ausencia (Rubin) encontraste en cada escenario y por qué tus estrategias de imputación no generaron sesgos severos.

## 2.3 Sensibilidad: Tratamiento de Outliers (Valores Atípicos)
Centra tu atención en la tabla de órdenes, específicamente en `days_since_prior_order` o conteos agrupados. Usando gráficos estadísticos (Boxplots) y computando el Rango Intercuartílico (IQR / Regla de Tukey) o Z-Scores, encuentra y reporta valores atípicos. ¿Existen usuarios que compran muy fuera del rango promedio de frecuencia?

### Conclusiones sobre la Sensibilidad:
Decide y justifica la estrategia que usaste para estos casos (eliminar, mantener como outlier real o winsorizar).

# Paso 3: Data Wrangling & Construcción de One Big Table (OBT)

Para poder responder preguntas de negocio de manera eficiente y aplicando correlaciones y cruces, un científico de datos típicamente convierte o "aplana" las tablas normalizadas operacionales en **One Big Table (OBT)**. La OBT agrupa (hace merges) de las distintas dimensiones transaccionales alrededor del concepto central (en este caso el *Articulo pedido por Orden*), lo que nos da una vista panorámica (desnormalización estructural) para el análisis.

**Tu tarea:** Une las cinco tablas previamente purgadas formando un solo DataFrame analítico. Te recomendamos ir uniendo ordenes y el detalle, luego producto, pasillo y finalmente departamento.

# Paso 4: Análisis Multivariado y de Negocio

Con tu conjunto de datos ahora limpio (*trusted/gold data*) mediante la auditoría y unificado en tu OBT, usa tus conocimientos de agregación y visualización estadística para responder las siguientes interrogantes. Usa histogramas, KDE y/o gráficos de dispersión donde la distribución lo merezca.

## [A] Preguntas Esenciales

**A1. La hora y el día: Verificación del dominio:**
Verifica con código que `order_hour_of_day` y `order_dow` tienen distribuciones lógicas basándonos en tu conocimiento del mundo real. Construye histogramas para ver qué picos de horas concentran compras. Aplica gráficos de barras para evaluar diferencias de días y horas (comparemos, por ejemplo, distribuciones de la demanda los Miércoles vs. Sábados). Verifica si existe una diferencia de horas por dia.

**A2. Distribuciones temporales de recompra:**
¿Cuánto tiempo transcurre estadísticamente para que alguien vuelva a realizar otra orden? Grafica este tiempo de espera y concluye sobre la concentración de los datos (¿asimetría?).

### Tira aquí tus hallazgos estadísticos para el grupo A:
(Debes hablar de cómo se distribuye la demanda, tendencias centrales encontradas y asimetrías de cola).

## [B] Profundización (Segmentación Categórica)

**B1. Retención y recurrencia (Número de órdenes por usuario):**
Aislando o agrupando a nivel de granularidad de cliente, describe la variabilidad y distribución de compras que hacen. ¿Existen colas largas de clientes extra-leales?

**B2. Productos Top: El principio de Pareto:**
Genera un top 20 de los productos más solicitados globalmente. Para cada producto reporta su ratio de recompra. ¿Algunos productos tienen una correlación fuerte entre ser de un 'pasillo' y volverse recompras aseguradas?

### Conclusiones sobre el perfil del carrito [Grupo B]:

## [C] Patrones de Causalidad y Exposición (Hard)

**C1. El tamaño de la canasta comercial:**
¿Cuántos artículos en promedio compran las personas estadísticamente? Aplica gráficos sobre esta proporción y analiza la curtosis (grado de acumulación) de esta distribución del tamaño del pedido.

**C2. ¿Son los primeros productos un síntoma de fidelidad? (Spearman o Pearson):**
Para los 20 objetos que con mayor frecuencia las personas agregan como posición #1 al carrito, elabora una lógica que calcule una matriz de correlación (en variables generadas o agrupadas) o genera gráficos de dispersión (Scatterplot). Compara si, a nivel de producto, ser posicionado primero tiene relación lineal con ser un producto de constante recompra global (el campo 'reordered').

**C3. Para cada producto, ¿cual es la proporcion de re-compras?**

**C4. Para cada usuario, ¿cual es la proporcion de re-compras de los productos comprados?**

**C5. ¿Cuales son los top 20 productos que los clientes ponen primero en sus carritos?**

### Tus conclusiones avanzadas [Grupo C]:
(Enuncia las posibles relaciones matemáticas encontradas o variables dispersas)

# Conclusión General Ejecutiva
Resume en un párrafo las principales fortalezas, falencias detectadas en la gobernanza inicial de este dataset, y los descubrimientos de negocio clave que proporcionarías a la directiva de Instacart.